_posts/2021-03-02-faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci.md

```diff
@@ -0,0 +1,279 @@
+        <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/00-cypress-files-before-knapsack.png" class="img-large" alt="List of Cypress test files, showing that tests are nested two levels: by module and by feature or area" />
+        <figcaption>List of Cypress test files, showing that tests are nested two levels: by module and by feature or area</figcaption>
+      </figure>
+    </a>
+  </article>
+  <article>
+    <a target="_blank" rel="noopener" href="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/01-pipeline-jobs-before-knapsack.png">
+      <figure>
+        <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/01-pipeline-jobs-before-knapsack.png" class="img-list" alt="List of Cypress test Gitlab CI jobs, showing that there are many separate jobs for each of the test subfolders" />
+        <figcaption>List of Cypress test Gitlab CI jobs, showing that there are many separate jobs for each of the test subfolders</figcaption>
+      </figure>
+    </a>
+  </article>
+  <article>
+    <a target="_blank" rel="noopener" href="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/02-cypress-runtime-before-knapsack.png">
+      <figure>
+        <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/02-cypress-runtime-before-knapsack.png" class="img-list" alt="List of Cypress test Gitlab CI jobs, showing that job runtimes vary from 53 seconds to 5 minutes" />
+        <figcaption>List of Cypress test Gitlab CI jobs, showing that job runtimes vary from 53 seconds to 5 minutes</figcaption>
+      </figure>
+    </a>
+  </article>
+  <article>
+    <a target="_blank" rel="noopener" href="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/03-pipeline-runtime-before-knapsack.png">
+      <figure>
+        <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/03-pipeline-runtime-before-knapsack.png" class="img-list" alt="Pipeline summary showing: 33 jobs for master in 19 minutes and 24 seconds" />
+        <figcaption>Pipeline summary</figcaption>
+      </figure>
+    </a>
+  </article>
+</section>
+
+# What is Knapsack Pro?
+
+Knapsack Pro is a tool that helps split up test suites in a dynamic way across parallel CI runners to ensure that each runner finishes work simultaneously.
+
+Its name comes from the “[knapsack problem](https://en.wikipedia.org/wiki/Knapsack_problem)”, a problem combinatorial optimization, to figure out how to take things (test suites) with certain attributes (number of tests, time it took on average to run, etc) and determine how best to divide them to get the most value (all CI nodes are running tests from boot until shutdown and they all finish at the same time).
```

Comment:
_a problem combinatorial optimization_ - is that part phrased correctly?

I know it would be correct this way: _a problem in combinatorial optimization_.

Unsure about the proposed version, so please confirm.

Reply:
I'll fix it, this is an obvious mistake =)